features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/java/org/opennms/features/topology/plugins/topo/linkd/internal/LinkdEdge.java

```diff
@@ -28,70 +28,266 @@
+            if (edge.getSpeed() == null && targetInterface.getIfSpeed() != null) {
+                edge.setSpeed(InetAddressUtils.getHumanReadableIfSpeed(targetInterface.getIfSpeed()));
+            }
+
+        }
+        
+        edge.setSourceAddr(sourceAddr);
+        edge.setTargetAddr(targetAddr);
+
+        return edge;
+    }
 
     private Integer m_sourceNodeid;
     private Integer m_targetNodeid;
 
-    private String m_sourceEndPoint;
-    private String m_targetEndPoint;
+    private String m_sourceLabel;
+    private String m_targetLabel;
+
+    private String m_sourceIfName;
+    private String m_targetIfName;
+
+    private String m_sourceAddr;
+    private String m_targetAddr;
+
+    private String m_speed;
+    
+    private Integer m_sourceIfIndex;
+    private Integer m_targetIfIndex;
+
+    private final ProtocolSupported m_discoveredBy;
     
-    public LinkdEdge(String namespace, String id, Vertex source, Vertex target) {
-        super(namespace, id, source, target);
+    private LinkdEdge(String id, Vertex source, Vertex target, ProtocolSupported discoveredBy) {
+        super(LinkdTopologyProvider.TOPOLOGY_NAMESPACE_LINKD, id, source, target);
+        m_discoveredBy = discoveredBy;
     }
 
-    public LinkdEdge(String namespace, String id, SimpleConnector source,
-            SimpleConnector target) {
-        super(namespace, id, source, target);
+    private LinkdEdge(String id, SimpleConnector source,
+            SimpleConnector target, ProtocolSupported discoveredBy) {
+        super(LinkdTopologyProvider.TOPOLOGY_NAMESPACE_LINKD, id, source, target);
+        m_discoveredBy = discoveredBy;
     }
 
-    public Integer getSourceNodeid() {
-        return m_sourceNodeid;
+    // Constructor to make cloneable easier for sub classes
+    private LinkdEdge(LinkdEdge edgeToClone) {
+            this(edgeToClone.getId(), edgeToClone.getSource().clone(), edgeToClone.getTarget().clone(),edgeToClone.getDiscoveredBy());
+            setSourceLabel(edgeToClone.getSourceLabel());
+            setSourceNodeid(edgeToClone.getSourceNodeid());
+            setSourceIfIndex(edgeToClone.getSourceIfIndex());
+            setSourceIfName(edgeToClone.getSourceIfName());
+            setSourceAddr(edgeToClone.getSourceAddr());
+            setTargetLabel(edgeToClone.getTargetLabel());
+            setTargetNodeid(edgeToClone.getTargetNodeid());
+            setTargetIfIndex(edgeToClone.getTargetIfIndex());
+            setTargetIfName(edgeToClone.getTargetIfName());
+            setTargetAddr(edgeToClone.getTargetAddr());
+            setSpeed(edgeToClone.getSpeed());
+            setLabel(edgeToClone.getLabel());
+            setStyleName(edgeToClone.getStyleName());
+            setTooltipText(edgeToClone.getTooltipText());
     }
 
-    public void setSourceNodeid(Integer sourceNodeid) {
-        m_sourceNodeid = sourceNodeid;
+    @Override
+    public LinkdEdge clone() {
```

Comment:
It is necessary because for some reasson the upper method getEdges() uses clone.
You get a copy of edges using clone....and if you want to be returned properly the LinkdEdge you must clone.